benchexec/resources.py

```diff
@@ -210,118 +218,738 @@ def _get_cpu_cores_per_run0(
+                        f"/sys/devices/system/cpu/cpu{core}/cache/{entry}/level"
                     )
-                if len(cores) >= coreLimit:
-                    break
-            cores_with_siblings.update(cores)
-            cores = cores[:coreLimit]  # shrink if we got more cores than necessary
-            # remove used cores such that we do not try to use them again
-            cores_of_unit[unit] = [
-                core for core in cores_of_unit[unit] if core not in cores
-            ]
-
-        assert len(cores) == coreLimit, (
-            f"Wrong number of cores for run {run + 1} of {num_of_threads} "
-            f"- previous results: {result}, "
-            f"remaining cores per CPU/memory region: {cores_of_unit}, "
-            f"current cores: {cores}"
+                )
+                == 3
+            ):
+                index_L3_cache = entry
+                break
+    """Get the id of the Level 3 cache a core belongs to."""
+    return int(
+        util.read_file(f"/sys/devices/system/cpu/cpu{core}/cache/{index_L3_cache}/id")
+    )
+
+
+def get_L3cache_mapping(allCpus_list: List[int]) -> HierarchyLevel:
+    cores_of_L3cache = collections.defaultdict(list)
+    try:
+        for core in allCpus_list:
+            L3cache = get_L3cache_id_for_core(core)
+            cores_of_L3cache[L3cache].append(core)
+    except FileNotFoundError:
+        cores_of_L3cache = {}
+        logging.error(
+            "Level 3 cache information not available at /sys/devices/system/cpu/cpuX/cache/cacheX"
         )
-        blocked_cores = cores if need_HT else cores_with_siblings
-        assert not used_cores.intersection(blocked_cores)
-        used_cores.update(blocked_cores)
-        result.append(sorted(cores))
-
-    assert len(result) == num_of_threads
-    assert all(len(cores) == coreLimit for cores in result)
-    assert (
-        len(set(itertools.chain(*result))) == num_of_threads * coreLimit
-    ), f"Cores are not uniquely assigned to runs: {result}"
-
-    logging.debug("Final core assignment: %s.", result)
-    return result
+    logging.debug("Level 3 caches of cores are %s.", cores_of_L3cache)
+    return cores_of_L3cache
+
+
+# returns dict of mapping NUMA region to list of cores
+def get_NUMA_mapping(allCpus_list: List[int]) -> HierarchyLevel:
+    cores_of_NUMA_region = collections.defaultdict(list)
+    for core in allCpus_list:
+        coreDir = f"/sys/devices/system/cpu/cpu{core}/"
+        NUMA_regions = _get_memory_banks_listed_in_dir(coreDir)
+        if NUMA_regions:
+            cores_of_NUMA_region[NUMA_regions[0]].append(core)
+            # adds core to value list at key [NUMA_region[0]]
+        else:
+            # If some cores do not have NUMA information, skip using it completely
+            logging.warning(
+                "Kernel does not have NUMA support. Use benchexec at your own risk."
+            )
+            cores_of_NUMA_region = {}
+            break
```

Comment:
It is easier to understand what happens if simply `return {}` is used instead. There is already a log message above, no reason for having another debug message with an empty set of cores.

Reply:
Done